CRM/Birthdays/Mailer.php

```diff
@@ -0,0 +1,121 @@
+        }
+
+        $email_id = Civi::settings()->get(CRM_Birthdays_Form_Settings::BIRTHDAYS_SENDER_EMAIL_ADDRESS_ID);
+        $this->email_address_from = $this->get_sender_email_address_from_id($email_id);
+        if (empty($this->email_address_from)) {
+            throw new Exception('Birthdays: Pre selected outgoing email not found. Please set am outgoing email address in birthday settings');
+        }
+    }
+
+    /**
+     * @param $contacts
+     * @param $write_activity
+     * @return int error count
+     * @throws API_Exception
+     * @throws CRM_Core_Exception
+     * @throws UnauthorizedException
+     */
+    public function send_mails_and_write_activity($contacts, $write_activity): int
+    {
+        $error_count = 0;
+        foreach ($contacts as $contact_id => $contact_info) {
+            try {
+                $this->send_mail($contact_id, $this->email_address_from, $contact_info['email'], $this->template_id);
+                if ($write_activity) $this->create_activity($contact_id, E::ts('Successful birthday greeting mail'), E::ts('Successful birthday greeting mail! Template ID %1 has been used.', [$this->template_id]));
+            } catch (Exception $exception) {
+                if ($write_activity) $this->create_activity($contact_id, E::ts('FAILED birthday greeting mail'), E::ts("Failed sending an birthday greeting mail with template ID nr %1. Error: %2", [$this->template_id, $exception]));
+                ++$error_count;
+            }
+        }
+        return $error_count;
+    }
+
+    /**
+     * @throws CRM_Core_Exception
+     * @throws Exception
+     */
+    private function send_mail($contact_id, $from_email_address, $to_email_address, $template_id): void
+    {
+        try {
+            civicrm_api3('MessageTemplate', 'send', [
+                'check_permissions' => 0,
+                'id'                => $template_id,
+                'to_name'           => civicrm_api3('Contact', 'getvalue', ['id' => $contact_id, 'return' => 'display_name']),
+                'from'              => trim($from_email_address),
+                'contact_id'        => $contact_id,
+                'to_email'          => trim($to_email_address),
+            ]);
+        } catch (Exception $exception) {
+            throw new Exception("Birthdays: MessageTemplate exception: $exception");
+        }
+    }
+
+
+    /**
+     * @param $target_id
+     * @param $title
+     * @param $description
+     * @return void
+     * @throws API_Exception
+     * @throws UnauthorizedException|CRM_Core_Exception
+     */
+    private function create_activity($target_id, $title, $description): void
+    {
+        Activity::create()
+            ->addValue('activity_type_id', 3) // = email
+            ->addValue('activity_type_id:label', 'E-Mail')
+            ->addValue('subject', E::ts($title))
+            ->addValue('details', $description)
+            ->addValue('source_contact_id', $target_id)
```

Comment:
Why set the source contact with the target contact ID? Shouldn't this be the currently logged-in user's contact? Wouldn't this be implicitly set if it wasn't provided?